upgrades/0.0.99.seperate-ns-from-app/pkg/migration/migrations_test.go

```diff
@@ -0,0 +1,409 @@
+package migration
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"path"
+	"strings"
+	"testing"
+	"text/template"
+
+	"github.com/oslokommune/okctl-upgrade/upgrades/0.0.99.seperate-ns-from-app/pkg/paths"
+
+	"github.com/oslokommune/okctl-upgrade/upgrades/0.0.99.seperate-ns-from-app/pkg/lib/manifest/apis/okctl.io/v1alpha1"
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+)
+
+func mockCluster(name string) v1alpha1.Cluster {
+	return v1alpha1.Cluster{
+		Metadata: v1alpha1.ClusterMeta{Name: name},
+		Github:   v1alpha1.ClusterGithub{OutputPath: "infrastructure"},
+	}
+}
+
+func contains(haystack []string, needle string) bool {
+	for _, item := range haystack {
+		if item == needle {
+			return true
+		}
+	}
+
+	return false
+}
+
+const appTemplate = `apiVersion: argoproj.io/v1alpha1
+kind: Application
+`
+
+func TestScanForRelevantApps(t *testing.T) {
+	testCases := []struct {
+		name        string
+		withCluster v1alpha1.Cluster
+		withFs      *afero.Afero
+		expectApps  []string
+	}{
+		{
+			name:        "Should return correct apps",
+			withCluster: mockCluster("mock-cluster"),
+			withFs: func() *afero.Afero {
+				fs := &afero.Afero{Fs: afero.NewMemMapFs()}
+
+				_ = fs.MkdirAll("/infrastructure/mock-cluster/argocd/applications", paths.DefaultFolderPermissions)
+				_ = fs.WriteReader("/infrastructure/mock-cluster/argocd/applications/mock-app-one.yaml", strings.NewReader(appTemplate))
+
+				return fs
+			}(),
+			expectApps: []string{"mock-app-one"},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			apps, err := getApplicationsInCluster(tc.withFs, tc.withCluster, "/")
+			assert.NoError(t, err)
+
+			assert.Equal(t, len(tc.expectApps), len(apps))
+
+			for _, item := range apps {
+				assert.True(t, contains(tc.expectApps, item))
+			}
+
+			for _, item := range tc.expectApps {
+				assert.True(t, contains(apps, item))
+			}
+		})
+	}
+}
+
+const namespaceTemplate = `apiVersion: v1
+kind: Namespace
+metadata:
+  name: {{.Name}}`
+
+func namespace(name string) io.Reader {
+	t := template.Must(template.New("namespace").Parse(namespaceTemplate))
+
+	buf := bytes.Buffer{}
+
+	_ = t.Execute(&buf, struct {
+		Name string
+	}{Name: name})
+
+	return &buf
+}
+
+func addOldAppNamespace(t *testing.T, fs *afero.Afero, appName string, namespaceName string) {
+	appBaseDir := path.Join("/infrastructure/applications", appName, "base")
+
+	err := fs.MkdirAll(appBaseDir, paths.DefaultFolderPermissions)
+	assert.NoError(t, err)
+
+	err = fs.WriteReader(path.Join(appBaseDir, "namespace.yaml"), namespace(namespaceName))
+	assert.NoError(t, err)
+}
+
+func TestMigrateApplication(t *testing.T) {
```

Comment:
I didn't think of this before today, but anyway, on possible change that might improve readability is this:

From the outside, what I really care about when running this upgrade, is to understand where were my files before, and where are they now. That is how I think I would be able to understand what happens best.

The test code could reflect this by doing somethng like this:

```golang
	testCases := []struct {
		// ...
		withFiles []string
		expectedFilesAfter []string
	}{
		{
			withFiles: []string{
				// Base folder
				// - Hello app
				"infrastructure/applications/hello/base/namespace.yaml",
				// Perhaps add more files to make the example realistic? Or perhaps it is just noise.

				// - Some other app
				"infrastructure/applications/some-other-app/base/namespace.yaml",

				// Cluster folder
				"infrastructure/my-cluster/argocd/applications/applications.yaml",

				// ... more files
			},
			expectedFilesAfter: []string{
				// Cluster folder
				"infrastructure/my-cluster/argocd/applications/namespaces.yaml",
				"infrastructure/my-cluster/argocd/namespaces/hello.yaml",
				"infrastructure/my-cluster/argocd/namespaces/some-other-namespace.yaml",
			},
		}
	}
```

Reply:
https://github.com/oslokommune/okctl-upgrade/pull/25/commits/5c5212cb985c93fcc1f0bd1ccb282cb4e37113d6

Reply:
Nice suggestion